src/coreclr/gc/gc.cpp

```diff
@@ -37862,26 +38329,74 @@ void gc_heap::decommit_ephemeral_segment_pages()
 
     gc_history_per_heap* current_gc_data_per_heap = get_gc_data_per_heap();
     current_gc_data_per_heap->extra_gen0_committed = heap_segment_committed (ephemeral_heap_segment) - heap_segment_allocated (ephemeral_heap_segment);
+#endif //MULTIPLE_HEAPS && USE_REGIONS
 }
 
-#ifdef MULTIPLE_HEAPS
 // return true if we actually decommitted anything
 bool gc_heap::decommit_step ()
 {
+    size_t decommit_size = 0;
+
+#ifdef USE_REGIONS
+    const size_t max_decommit_step_size = DECOMMIT_SIZE_PER_MILLISECOND * DECOMMIT_TIME_STEP_MILLISECONDS;
+    for (int kind = basic_free_region; kind < count_free_region_kinds; kind++)
+    {
+        dprintf (REGIONS_LOG, ("decommit_step %d, regions_to_decommit = %Id",
+            kind, global_regions_to_decommit[kind].get_num_free_regions()));
+        while (global_regions_to_decommit[kind].get_num_free_regions() > 0)
+        {
+            heap_segment* region = global_regions_to_decommit[kind].unlink_region_front();
+
+            uint8_t* page_start = align_lower_page(get_region_start(region));
+            uint8_t* end = use_large_pages_p ? heap_segment_used(region) : heap_segment_committed(region);
+            size_t size = end - page_start;
+            bool decommit_succeeded_p = false;
+            if (!use_large_pages_p)
+            {
+                decommit_succeeded_p = virtual_decommit(page_start, size, heap_segment_oh(region), 0);
+                dprintf(REGIONS_LOG, ("decommitted region %Ix(%Ix-%Ix) (%Iu bytes) - success: %d",
+                    region,
+                    page_start,
+                    end,
+                    size,
+                    decommit_succeeded_p));
+            }
+            if (!decommit_succeeded_p)
+            {
+                memclr(page_start, size);
```

Comment:
if we are going to delete the region in region allocator do we need to memclr here? any region allocated from there will be reinitialized.

Reply:
I think we do, because when we get the region from the region allocator, the memory will be committed, but if it's already committed, any garbage values from previous objects would just stay there.

Reply:
in that case you could set heap_segment_used to heap_segment_committed when you get a new region on large pages. and the allocator should do the right thing.

Reply:
it looks like setting heap_segment_used wouldn't actually do anything because it gets reinitialized in make_heap_segment called from allocate_new_region

Reply:
yes, you can avoid doing that in `make_heap_segment` which already has special handling for large pages (I changed how used is assigned) -

```cplusplus
    heap_segment_reserved (new_segment) = new_pages + size;
    heap_segment_committed (new_segment) = (use_large_pages_p ? 
        heap_segment_reserved(new_segment) : (new_pages + initial_commit));
    heap_segment_used (new_segment) = (use_large_pages_p ? heap_segment_committed (new_segment) : start);
```